switch from `training_args.bin` `training_args.json`

not-lain opened the pull request:
# What does this PR do?

switch from  `training_args.bin` to `training_args.json` and only capture the parameters that the user passed 
I'm using the same approach we are using in huggingface_hub's [PyTorchModelHubMixin](https://github.com/huggingface/huggingface_hub/blob/503d35355fa79ede904665c944e0a247e85c8af7/src/huggingface_hub/hub_mixin.py#L277) to store as little parameters as possible.
a minimalistic approach to test this is pr

```py
from transformers import TrainingArguments
args = TrainingArguments(output_dir="folder",eval_strategy="no") # or any other paramters
print(args.to_json_string())
# outputs
"""
{
  "output_dir": "folder",
  "eval_strategy": "no",
  "logging_dir": "folder\\runs\\Nov29_02-44-45_iphone-laptop"
}
"""
# logging_dir is a special parameter that is always captured and added to the training_args because we want to ensure consistency

# stores the parameters into a file
args.to_json_file("training_args.json")

# loads an instance using the class directly
args2  = TrainingArguments.from_json_file("training_args.json") 
```

using this approach, we ensure that we only store the parameters that the user-defined manually and not the ones that have default values or the ones inferred from the system (ie cpu, cuda, tpu ... ), leaving some room for flexibility.

in a sense the parameters are mutable, meaning the user can physically alter them.


Fixes #34612 


## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?


## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

@muellerzr @SunMarc

not-lain committed:
capture init parameters in training_args

not-lain committed:
update relevant attributes

not-lain committed:
attribute calling for training args

not-lain committed:
add class attribute to load the training_args from a local file

not-lain committed:
Merge branch 'main' into switch-training_args-file-format

not-lain commented on src/transformers/training_args.py:
```diff
@@ -3104,3 +3128,21 @@ class ParallelMode(Enum):
+
+def serialize_parameter(k, v):
+    if k == "torch_dtype" and not isinstance(v, str):
+        return str(k).split(".")[1]
```
```suggestion
        return str(v).split(".")[1]
```

not-lain committed:
Update src/transformers/training_args.py

not-lain commented:
cc @muellerzr & @SunMarc for review.
failed test is unrelated

not-lain committed:
Merge branch 'main' into switch-training_args-file-format

not-lain commented:
friendly tagging @muellerzr & @SunMarc
failed tests are unrelated

muellerzr reviewed (commented):
Very nice! I'm a fan of this change. However, we should be careful about old training resumptions from older versions and likely should do a deprecation until 5.0.0 (warning *if* a `.bin` was found but we do the shift to `.json` now)

How does that sound?

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/transformers/pr_35010). All of your documentation changes will be reflected on that endpoint. The docs are available until 30 days after the last update.

not-lain committed:
Merge branch 'huggingface:main' into switch-training_args-file-format

not-lain committed:
ensure backward compatibility

not-lain committed:
fix version parsing

not-lain committed:
fix deprecated filename

not-lain committed:
update test

not-lain committed:
format with ruff

not-lain committed:
fix trainer logger tests

not-lain commented:
[don't merge yet I need to fix something fundamental that was not captured by the tests]

not-lain committed:
serialize all parameters

not-lain commented:
cc @muellerzr for feedback
I implemented a rollback, now default is `traing_args.bin` and people can switch to `training_args.json` by setting the `TRAINER_SAFE_SERIALIZE` variable.

originally there was never a mechanism inside of the trainer API to load any training_args and people used to load the training arguments manually, I added something useful which is `TrainingArguments.from_json_file(file_name )` 
but I think I will integrate this loading mechanism with the trainer api in another pr (optional new feature)

not-lain commented:
friendly tagging @muellerzr here for a review.

not-lain commented:
cc @muellerzr for review

SunMarc reviewed (commented):
Thanks for the PR ! Left a few comments !

not-lain commented on src/transformers/trainer.py:
```diff
@@ -301,14 +301,21 @@ def safe_globals():
 # Name of the files used for checkpointing
-TRAINING_ARGS_NAME = "training_args.bin"
+TRAINING_ARGS_NAME = "training_args.json"
+DEPRECATED_ARGS_NAME = "trainer_state.bin"
```
nice catch

not-lain committed:
Update src/transformers/trainer.py

Co-authored-by: Marc Sun <57196510+SunMarc@users.noreply.github.com>

not-lain committed:
Refactor serialization logic in Trainer to use binary serialization flag

not-lain committed:
Merge branch 'main' into switch-training_args-file-format

not-lain committed:
format with ruff

not-lain committed:
Merge branch 'switch-training_args-file-format' of https://github.com/not-lain/transformers into switch-training_args-file-format

not-lain committed:
switch to dynamic serialization

not-lain committed:
Refactor training_args serialization to only include defined dataclass fields

not-lain committed:
ensure consistant traininglogs to ensure that tensorboard logs continue from where they left off

not-lain commented:
cc @SunMarc for review

I've switched to dynamic serialization which something that came up with to tackle #25903 .
this will serialize the parameters needed to make the training arguments **lighweight, mutable and reproducible.**

here's a thorough example on what to expect : 
```py
from transformers import TrainingArguments
args = TrainingArguments(output_dir="folder",eval_strategy="no") # or any other parameters
print(args.to_json_string())
# outputs (logging_dir needs to stay consistant in case we want tensorboard to continue the plots)
"""
{
  "output_dir": "folder",
  "eval_strategy": "no",
  "logging_dir": "folder\\runs\\Dec27_21-48-40_iphone-laptop"
}
"""
# has some default parameters in the method which have been captured and serialized
# this also captures manually set parameters in either args or kwargs format.
args = args.set_dataloader() 
print(args.to_json_string())
# outputs
"""
{
  "output_dir": "folder",
  "eval_strategy": "no",
  "logging_dir": "folder\\runs\\Dec27_21-48-40_iphone-laptop",
  "auto_find_batch_size": false,
  "ignore_data_skip": false
}
"""
# save and reload
args.to_json_file("training_args.json")
args2  = TrainingArguments.from_json_file("training_args.json")
print(args2.to_json_string())
# output
"""
{
  "output_dir": "folder",
  "eval_strategy": "no",
  "logging_dir": "folder\\runs\\Dec27_21-48-40_iphone-laptop",
  "auto_find_batch_size": false,
  "ignore_data_skip": false
}
"""
```

let me know if you have any reviews

SunMarc reviewed (approved):
LGTM ! thanks for iterating ! Can you have a second look @muellerzr ?

SunMarc commented on src/transformers/training_args.py:
```diff
@@ -2529,25 +2530,33 @@ def to_dict(self):
-            if is_accelerate_available() and isinstance(v, AcceleratorConfig):
-                d[k] = v.to_dict()
+            d[k] = serialize_parameter(k, v)
         self._dict_torch_dtype_to_str(d)
```
Looks fine thanks ! SInce we are not using `_dict_torch_dtype_to_str` method, you can also remove it in this PR

muellerzr commented on src/transformers/trainer.py:
```diff
@@ -3938,7 +3949,14 @@ def _save(self, output_dir: Optional[str] = None, state_dict=None):
-        torch.save(self.args, os.path.join(output_dir, TRAINING_ARGS_NAME))
+        if not binary_serializiation:
+            self.args.to_json_file(os.path.join(output_dir, TRAINING_ARGS_NAME))
+            logger.info(
+                "safe serialization has been enabled by default and the training args will be stored in a {TRAINING_ARGS_NAME} \n"
+                "to fall back to the {DEPRECATED_ARGS_NAME} you can set os.environ['binary_serializiation']= '1' "
+            )
```
Let's be a bit more explicit. Do a `warnings.warn()` & mention deprecation in 5.0.0 here please! :)

muellerzr commented on src/transformers/training_args.py:
```diff
@@ -67,6 +69,25 @@
+T = TypeVar("T", bound="TrainingArguments")
+
+
+def serialize(func):
```
Small docstring please :)

not-lain replied:
```suggestion
def serialize(func):
	"""
	A decorator that captures and serializes the parameters of any method that was called from the original class and stores all valid parameters in the `__training_args_params__` attribute.
	Args: 
		func: The function to be decorated.
	Returns:
		The wrapped function with serialized parameters.
	"""
```

muellerzr replied:
Sounds perfect!

muellerzr commented on src/transformers/training_args.py:
```diff
@@ -3090,3 +3143,21 @@ class ParallelMode(Enum):
     TPU = "tpu"
+
+
+def serialize_parameter(k, v):
```
Docstring here too

not-lain replied:
```suggestion
def serialize_parameter(k, v):
	"""
	Serializes a parameter based on its type and key.
	This function takes a key and value, and serializes the value depending on its type.
	Args: 
		k: The key associated with the parameter. 
		v: The value to be serialized, which can be of various types.
	Returns:: 
		The serialized value, which depends on its original type and key.
	"""
```

muellerzr reviewed (commented):
Thanks so much @not-lain ! So sorry it took me so long. 

This looks great, just a few nits for docstrings + let's deprecate out to 5.0 and warn users please!

not-lain committed:
Update src/transformers/training_args.py

not-lain committed:
Update src/transformers/training_args.py

not-lain committed:
Merge branch 'main' into switch-training_args-file-format

not-lain committed:
update with ruff

not-lain committed:
switch from logging info to logging warning

not-lain committed:
Change logger from info to warning for safe serialization message and update test assertion for warning capture

not-lain committed:
debugging stateful callback

not-lain committed:
update test

not-lain committed:
Merge branch 'main' into switch-training_args-file-format

not-lain committed:
thanks claude

not-lain committed:
Merge branch 'switch-training_args-file-format' of https://github.com/not-lain/transformers into switch-training_args-file-format

not-lain commented:
cc @muellerzr , failing tests are unrelated

not-lain committed:
Merge branch 'main' into switch-training_args-file-format

not-lain committed:
Merge branch 'main' into switch-training_args-file-format

not-lain commented:
friendly tagging @muellerzr  for review

not-lain commented:
cc @muellerzr for review

not-lain commented:
cc @muellerzr to see if you can review this

muellerzr commented on src/transformers/trainer.py:
```diff
@@ -4575,7 +4593,14 @@ def _push_from_checkpoint(self, checkpoint_folder):
+            self.args.to_json_file(os.path.join(output_dir, TRAINING_ARGS_NAME))
+            logger.warn(
+                "Safe serialization has been enabled by default and the training args will be stored in a {TRAINING_ARGS_NAME} \n"
+                "to fall back to the {DEPRECATED_ARGS_NAME} you can set os.environ['binary_serializiation']= '1', this feature will be removed in version 5.0.0"
```
```suggestion
                f"to fall back to the {DEPRECATED_ARGS_NAME} you can set os.environ['binary_serializiation']= '1', this feature will be removed in version 5.0.0"
```

muellerzr reviewed (approved):
Thanks a bunch! fstring nit then passing off to @ArthurZucker :)

(Merged myself though so we can get this in without more delay!)

muellerzr committed:
Update src/transformers/trainer.py

not-lain commented on src/transformers/trainer.py:
```diff
@@ -3768,7 +3772,14 @@ def _save_tpu(self, output_dir: Optional[str] = None):
+            if not binary_serializiation:
+                self.args.to_json_file(os.path.join(output_dir, TRAINING_ARGS_NAME))
+                logger.warn(
+                    "Safe serialization has been enabled by default and the training args will be stored in a {TRAINING_ARGS_NAME} \n"
+                    "to fall back to the {DEPRECATED_ARGS_NAME} you can set os.environ['binary_serializiation']= '1', this feature will be removed in version 5.0.0"
```
```suggestion
                    f"Safe serialization has been enabled by default and the training args will be stored in a {TRAINING_ARGS_NAME} \n"
                    f"to fall back to the {DEPRECATED_ARGS_NAME} you can set os.environ['binary_serializiation']= '1', this feature will be removed in version 5.0.0"
```

not-lain commented on src/transformers/trainer.py:
```diff
@@ -3865,7 +3876,14 @@ def _save(self, output_dir: Optional[str] = None, state_dict=None):
+        if not binary_serializiation:
+            self.args.to_json_file(os.path.join(output_dir, TRAINING_ARGS_NAME))
+            logger.warn(
+                "Safe serialization has been enabled by default and the training args will be stored in a {TRAINING_ARGS_NAME} \n"
+                "to fall back to the {DEPRECATED_ARGS_NAME} you can set os.environ['binary_serializiation']= '1', this feature will be removed in version 5.0.0"
```
```suggestion
                f"Safe serialization has been enabled by default and the training args will be stored in a {TRAINING_ARGS_NAME} \n"
                f"to fall back to the {DEPRECATED_ARGS_NAME} you can set os.environ['binary_serializiation']= '1', this feature will be removed in version 5.0.0"
```

not-lain commented on src/transformers/trainer.py:
```diff
@@ -4575,7 +4593,14 @@ def _push_from_checkpoint(self, checkpoint_folder):
+        if not binary_serializiation:
+            self.args.to_json_file(os.path.join(output_dir, TRAINING_ARGS_NAME))
+            logger.warn(
+                "Safe serialization has been enabled by default and the training args will be stored in a {TRAINING_ARGS_NAME} \n"
```
```suggestion
                f"Safe serialization has been enabled by default and the training args will be stored in a {TRAINING_ARGS_NAME} \n"
```

not-lain committed:
Update src/transformers/trainer.py

not-lain committed:
Update src/transformers/trainer.py

not-lain committed:
Update src/transformers/trainer.py

not-lain committed:
Merge branch 'main' into switch-training_args-file-format

ArthurZucker reviewed (approved):
I have no idea where this stands, happy to merge

not-lain commented:
this pr is related to https://github.com/huggingface/transformers/pull/35010/files#diff-bfceaff300c851b8e24fc50dc6638482abaec8f7d2a718e877c3828c166bcf79L223 where  it said  : 
>  TODO: `TrainingArguments` users rely on it being fully mutable. In the future see if we can narrow this to a few keys: https://github.com/huggingface/transformers/pull/25903

in short this pr : 
* switch from bin -> json (better security) 
* training args become fully mutable
* dynamic parameter serialization (storing only key parameters) (see  https://github.com/huggingface/transformers/pull/35010#issuecomment-2564059045  for more details ) 

do not hesitate if you have any questions or follow-ups 🤗

not-lain committed:
resolve merge conflict

not-lain committed:
Merge branch 'main' into switch-training_args-file-format

not-lain commented:
done (≧∇≦)ﾉ

SunMarc commented on src/transformers/training_args.py:
```diff
@@ -1549,6 +1577,19 @@ class TrainingArguments:
         },
     )
 
+    def __new__(self, *args, **kwargs):
+        # catch and save only the parameters that the user passed
+        self.__training_args_params__ = {}
+        param_names = list(self.__dataclass_fields__.keys())
+
+        for i in range(len(args)):
+            self.__training_args_params__[param_names[i]] = serialize_parameter(param_names[i], args[i])
+
+        for k, v in kwargs.items():
+            self.__training_args_params__[k] = serialize_parameter(k, v)
+
+        return super().__new__(self)
+
```
just a caveat of only storing training params that were passed, if we change the default of any argument in the future, it will impact the training of the model if the user try to resume the training no ? it is a pretty small edge case but still nice to be aware of it.
Also, does this happen with the bin file ?

not-lain replied:
I can't deny the fact that when we change the default params the default training arguments will change, but still what if we remove some of the default parameters, the approach introduced in this pr is still more flexible (note that the _transformers version_ and some of these parameters are already part of the [readme file](https://huggingface.co/edbeeching/test-trainer-to-hub) so that the user can always have a fallback if they want to ) 

another good example of how this shows more prospects is that if we change the training device, if the second *server* supports more default gpus then that will be default parameter over there.

In full transparency, in bin, everything is static, and i haven't seen anyone report any problems with it.

qgallouedec replied:
Thanks @SunMarc, no change required on TRL side 👌
